src/GafferUI/GraphGadget.cpp

```diff
@@ -756,8 +1003,183 @@ ConnectionGadget *GraphGadget::reconnectionGadgetAt( const NodeGadget *gadget, c
 	return nullptr;
 }
 
+void GraphGadget::dirtyActive()
+{
+	if( m_activeStateTask )
+	{
+		m_activeStateTask->cancelAndWait();
+		m_activeStateTask.reset();
+	}
+	m_activeStateDirty = true;
+	dirty( DirtyType::Render );
+}
+
+void GraphGadget::updateActive() const
+{
+	if( m_activeStateTask )
+	{
+		m_activeStateTask->cancelAndWait();
+		m_activeStateTask.reset();
+	}
+
+	tbb::spin_mutex::scoped_lock scopedLockA( m_pendingActiveStateMutex );
+	const Gaffer::Node *focusNode = m_scriptNode->getFocus();
+	std::vector<Gaffer::ConstPlugPtr> focusPlugs;
+	if( focusNode )
+	{
+		Gaffer::ConstPlugPtr hiddenFocusPlug = nullptr;
+		for( auto &plug : Gaffer::Plug::OutputRange( *focusNode ) )
+		{
+			const std::string &n = plug->getName();
+			if( n.size() >= 2 && n[0] == '_' && n[1] == '_' )
+			{
+				if( !hiddenFocusPlug )
+				{
+					hiddenFocusPlug = plug;
+				}
+				continue;
+			}
+
+			focusPlugs.push_back( plug );
+		}
+
+		if( !focusPlugs.size() && hiddenFocusPlug )
+		{
+			// If we couldn't find any visible output plug, take the hidden one
+			focusPlugs.push_back( hiddenFocusPlug );
+		}
+	}
+
+	if( !focusPlugs.size() )
+	{
+		m_pendingActiveState.reset();
+		scopedLockA.release();
+		Gaffer::ParallelAlgo::callOnUIThread(
+			[this] {
+				this->applyActive( false );
+			}
+		);
+		return;
+	}
+
+
+	tbb::spin_mutex &pendingActiveStateMutex = m_pendingActiveStateMutex;
+	std::unique_ptr<PendingActive> &pendingActiveState = m_pendingActiveState;
+
+	// Must hold a reference to stop us dying before our UI thread call is scheduled.
+	ConstGraphGadgetPtr thisRef = this;
+
+	m_activeStateTask = Gaffer::ParallelAlgo::callOnBackgroundThread(
+		// Subject:  TODO - does it matter that we're only including one plug as a subject?  We respond to any
```

Comment:
It definitely doesn't matter right now, because cancellation for graph edits is done at the level of the ScriptNode (any edit anywhere in the script will cancel us). I don't know if we'll ever get to doing that at a finer level of granularity, but it does seem that `callOnBackgroundThread()` needs a more flexible way of specifying a subject (or subjects). See related comments in GraphComponentPath in my recent PR.

Not something we should worry about now, but it does add to the evidence that the signature for `callOnBackgroundThread()` isn't quite right.

Reply:
Clarified comment a bit in 078a377